src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +274,527 @@ create_distributed_table(PG_FUNCTION_ARGS)
+ * keys for the given table are supported.
+ *
+ * We allow distributed -> reference
+ *          distributed -> citus local
+ *
+ * We disallow reference   -> distributed
+ *             citus local -> distributed
+ *             regular     -> distributed
+ *
+ * Normally regular		-> distributed is allowed but it is not allowed when we create the
+ * distributed table concurrently because we rely on conversion of regular table to citus local table,
+ * which errors with an unfriendly message.
+ */
+static void
+EnsureForeignKeysForDistributedTableConcurrently(Oid relationId)
+{
+	/*
+	 * disallow citus local -> distributed fkeys.
+	 * disallow reference   -> distributed fkeys.
+	 * disallow regular     -> distributed fkeys.
+	 */
+	EnsureNoFKeyFromTableType(relationId, INCLUDE_CITUS_LOCAL_TABLES |
+							  INCLUDE_REFERENCE_TABLES | INCLUDE_LOCAL_TABLES);
+
+	/*
+	 * disallow distributed -> regular fkeys.
+	 */
+	EnsureNoFKeyToTableType(relationId, INCLUDE_LOCAL_TABLES);
+}
+
+
+/*
+ * EnsureColocatedWithTable ensures given relation can be colocated with the table of given name.
+ */
+static void
+EnsureColocatedWithTable(Oid relationId, char distributionMethod,
+						 char *distributionColumnName, char *colocateWithTableName)
+{
+	bool viaDeprecatedAPI = false;
+	char replicationModel = DecideReplicationModel(distributionMethod,
+												   colocateWithTableName,
+												   viaDeprecatedAPI);
+
+	/*
+	 * we fail transaction before local table conversion if the table could not be colocated with
+	 * given table. We should make those checks after local table conversion by acquiring locks to
+	 * the relation because the distribution column can be modified in that period.
+	 */
+	Relation relation = relation_open(relationId, AccessShareLock);
+
+	Var *distributionColumn = BuildDistributionKeyFromColumnName(relationId,
+																 distributionColumnName,
+																 NoLock);
+	Oid distributionColumnType = distributionColumn->vartype;
+
+	text *colocateWithTableNameText = cstring_to_text(colocateWithTableName);
+	Oid colocateWithTableId = ResolveRelationId(colocateWithTableNameText, false);
+	EnsureTableCanBeColocatedWith(relationId, replicationModel,
+								  distributionColumnType, colocateWithTableId);
+
+	relation_close(relation, AccessShareLock);
+}
+
+
+/*
+ * AcquireCreateDistributedTableConcurrentlyLock does not allow concurrent create_distributed_table_concurrently
+ * operations.
+ */
+void
+AcquireCreateDistributedTableConcurrentlyLock(Oid relationId)
+{
+	LOCKTAG tag;
+	SET_LOCKTAG_CREATE_DISTRIBUTED_TABLE_CONCURRENTLY(tag);
+	const bool sessionLock = false;
+	const bool dontWait = true;
+
+	LockAcquireResult lockAcquired = LockAcquire(&tag, ExclusiveLock, sessionLock,
+												 dontWait);
+	if (!lockAcquired)
+	{
+		ereport(ERROR, (errmsg(
+							"could not acquire the lock required to create distributed "
+							"table concurrently %s.", generate_qualified_relation_name(
```

Comment:
maybe: `"another create_distributed_table_concurrently operation is in progress"`

it's better to get straight to the point in the error message, because users do not always see detail messages in their application logs